ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -1148,26 +1152,60 @@ public List<AmbryReplica> getReplicaIdsByState(AmbryPartition partition, Replica
       Set<AmbryReplica> replicas = new HashSet<>();
       List<String> dcs = dcName != null ? Collections.singletonList(dcName)
           : dcToDcInfo.values().stream().map(dcInfo -> dcInfo.dcName).collect(Collectors.toList());
+      String partitionName = partition.toPathString();
       for (String dc : dcs) {
-        String resourceName = partitionToResourceNameByDc.get(dc).get(partition.toPathString());
+        int replicaSizeBefore = replicas.size();
+        String resourceName = partitionToResourceNameByDc.get(dc).get(partitionName);
         RoutingTableSnapshot routingTableSnapshot =
             clusterMapConfig.clusterMapUseAggregatedView ? globalRoutingTableSnapshotRef.get()
                 : dcToRoutingTableSnapshotRef.get(dc).get();
-        routingTableSnapshot.getInstancesForResource(resourceName, partition.toPathString(), state.name())
-            .stream()
-            .map(instanceConfig -> instanceNameToAmbryDataNode.get(instanceConfig.getInstanceName()))
-            .filter(dataNode -> dataNode.getDatacenterName().equals(dc))
-            .map(dataNode -> ambryDataNodeToAmbryReplicas.get(dataNode).get(partition.toPathString()))
-            .filter(Objects::nonNull)
-            .forEach(replicas::add);
+        getReplicaIdsByStateInRoutingTableSnapshot(routingTableSnapshot, dc, resourceName, partitionName, state,
+            replicas);
+        if (replicas.size() != replicaSizeBefore) {
```

Comment:
Should this be `replicas.size() > replicaSizeBefore` to be more precise?

Reply:
we are only adding to the list, it should be fine.